### PR TITLE
SONAR-10365 Use pull request id as key for pull requests

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 1]
```diff
@@ -112,13 +113,12 @@ private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder re
     Optional<BranchDto> mergeBranch = Optional.ofNullable(mergeBranchesByUuid.get(branch.getMergeBranchUuid()));
     ProjectPullRequests.PullRequest.Builder builder = ProjectPullRequests.PullRequest.newBuilder();
     String branchKey = branch.getKey();
+
+    DbProjectBranches.PullRequestData pullRequestData = branch.getPullRequestData();
     builder.setId(branchKey);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG35kc9JosG6VNHvO3C&open=AWG35kc9JosG6VNHvO3C)

[ISSUE_KEY]: AWG35kc9JosG6VNHvO3C

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 2]
```diff
@@ -112,13 +113,12 @@ private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder re
     Optional<BranchDto> mergeBranch = Optional.ofNullable(mergeBranchesByUuid.get(branch.getMergeBranchUuid()));
     ProjectPullRequests.PullRequest.Builder builder = ProjectPullRequests.PullRequest.newBuilder();
     String branchKey = branch.getKey();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Bug Bug: A "NullPointerException" could be thrown; "getPullRequestData()" can return null. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG35kc9JosG6VNHvO3E&open=AWG35kc9JosG6VNHvO3E)

[ISSUE_KEY]: AWG35kc9JosG6VNHvO3E

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 3]
```diff
@@ -112,13 +113,12 @@ private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder re
...
+    DbProjectBranches.PullRequestData pullRequestData = branch.getPullRequestData();
     builder.setId(branchKey);
-    //TODO get the branch from PULL_REQUEST_DATA
-    builder.setBranch(branchKey);
-    //TODO get the URL from PULL_REQUEST_DATA
-    builder.setUrl("https://github.com/SonarSource/sonar-core-plugins/pull/32");
-    //TODO get the branch title from PULL_REQUEST_DATA
-    builder.setTitle(branchKey);
+    builder.setBranch(pullRequestData.getBranch());
+    builder.setUrl(pullRequestData.getUrl());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG35kc9JosG6VNHvO3D&open=AWG35kc9JosG6VNHvO3D)

[ISSUE_KEY]: AWG35kc9JosG6VNHvO3D

File: server/sonar-server/src/main/resources/org/sonar/server/branch/pr/ws/list-example.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "pullRequests": [
     {
-      "id": "feature/bar",
+      "id": "123",
       "title": "feature/bar",
```
Comment (henryju) [inline server/sonar-server/src/main/resources/org/sonar/server/branch/pr/ws/list-example.json]: Shouldn't you change the sample title to make it clear this is maybe not equal to the branch name?

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 5]
```diff
@@ -112,13 +113,12 @@ private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder re
...
+
+    DbProjectBranches.PullRequestData pullRequestData = branch.getPullRequestData();
     builder.setId(branchKey);
-    //TODO get the branch from PULL_REQUEST_DATA
-    builder.setBranch(branchKey);
-    //TODO get the URL from PULL_REQUEST_DATA
-    builder.setUrl("https://github.com/SonarSource/sonar-core-plugins/pull/32");
-    //TODO get the branch title from PULL_REQUEST_DATA
-    builder.setTitle(branchKey);
+    builder.setBranch(pullRequestData.getBranch());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Bug Bug: A "NullPointerException" could be thrown; "pullRequestData" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG38FNK42pHCPIO0Eiw&open=AWG38FNK42pHCPIO0Eiw)

[ISSUE_KEY]: AWG38FNK42pHCPIO0Eiw

File: server/sonar-server/src/main/resources/org/sonar/server/branch/pr/ws/list-example.json (hunk 4 above)
Comment (janos-ss) [reply server/sonar-server/src/main/resources/org/sonar/server/branch/pr/ws/list-example.json]: Good idea, thanks!

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 6]
```diff
@@ -110,15 +111,16 @@ public void handle(Request request, Response response) throws Exception {
...
-    builder.setBranch(branchKey);
-    //TODO get the URL from PULL_REQUEST_DATA
-    builder.setUrl("https://github.com/SonarSource/sonar-core-plugins/pull/32");
-    //TODO get the branch title from PULL_REQUEST_DATA
-    builder.setTitle(branchKey);
+    builder.setId(branch.getKey());
+
+    // known to be non-null; all pull requests have non-null value, and caller filters accordingly
+    DbProjectBranches.PullRequestData pullRequestData = branch.getPullRequestData();
+    builder.setBranch(pullRequestData.getBranch());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Bug Bug: A "NullPointerException" could be thrown; "pullRequestData" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG38FNK42pHCPIO0Eiw&open=AWG38FNK42pHCPIO0Eiw)

[ISSUE_KEY]: AWG38FNK42pHCPIO0Eiw

Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Bug Bug: A "NullPointerException" could be thrown; "pullRequestData" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjanos%2FSONAR-10365%2Fuse-pr-id-as-key-for-pull-requests&issues=AWG38FNK42pHCPIO0Eiw&open=AWG38FNK42pHCPIO0Eiw)

[ISSUE_KEY]: AWG38FNK42pHCPIO0Eiw

**Pull request merged**